source/particle/world.cc

```diff
@@ -852,15 +970,199 @@ namespace aspect
+                                                                  update_flags,
+                                                                  component_indices[composition]));
+
+            if (simulator_access.include_melt_transport())
+              {
+                melt_component_indices[0] = simulator_access.introspection().variable("fluid velocity").first_component_index;
+                melt_component_indices[1] = simulator_access.introspection().variable("fluid pressure").first_component_index;
+                melt_component_indices[2] = simulator_access.introspection().variable("compaction pressure").first_component_index;
+
+                fluid_velocity = std::make_unique<FEPointEvaluation<dim, dim>>(simulator_access.get_mapping(),
+                                                                               simulator_access.get_fe(),
+                                                                               update_flags,
+                                                                               melt_component_indices[0]);
+                fluid_pressure = std::make_unique<FEPointEvaluation<1, dim>>(simulator_access.get_mapping(),
+                                                                             simulator_access.get_fe(),
+                                                                             update_flags,
+                                                                             melt_component_indices[1]);
+                compaction_pressure = std::make_unique<FEPointEvaluation<1, dim>>(simulator_access.get_mapping(),
+                                                                                  simulator_access.get_fe(),
+                                                                                  update_flags,
+                                                                                  melt_component_indices[2]);
+
+              }
+          }
+
+          void
+          reinit(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                 const ArrayView<Point<dim>> &positions,
+                 const ArrayView<double> &solution_values,
+                 const UpdateFlags update_flags)
+          {
+            EvaluationFlags::EvaluationFlags evaluation_flags = EvaluationFlags::nothing;
+
+            if (update_flags & update_values)
+              evaluation_flags = evaluation_flags | EvaluationFlags::values;
+
+            if (update_flags & update_gradients)
+              evaluation_flags = evaluation_flags | EvaluationFlags::gradients;
+
+            Assert ((update_flags & ~(update_gradients | update_values)) == false,
+                    ExcNotImplemented());
+
+            velocity.reinit (cell, positions);
+            pressure.reinit (cell, positions);
+            temperature.reinit (cell, positions);
+
+            for (auto &evaluator_composition: compositions)
+              evaluator_composition.reinit (cell, positions);
```

Comment:
Are these expensive calls? If so, it might actually be nice to run all of these `reinit()` calls in parallel using `Threads::new_task()`, but maybe that's overkill. It also only helps with serial runs.

Not necessary to change, just trying to plant a seed :-)

Reply:
I tend to think of ASPECT as a purely MPI parallel code with a very occasional use of threads for writing files. I know that the assembly uses the workstream interface, but other than that I don't think I usually think about parallelizing things in threads. Maybe it would be an interesting project at some point to investigate the potential of more threading, but until then I prefer the simpler code and concepts to a potential speedup for a small set of models. :-)

Reply:
Yes, and when we use MPI, we do actually disable threading as well. But I think that a lot of initial tests are run without MPI and where it's easy, it might actually be nice to use threading.

But it's not a priority any of us should have :-)